### PR TITLE
add numa to improve cpu inference perf

### PR DESCRIPTION
**Pull request by sywangyi**
@OlivierDehaene OR @Narsil

### CODE REVIEW

**Comment by sywangyi** [comment]
15% improvement in time_per_token in Intel(R) Xeon(R) Platinum 8480L

**Comment by sywangyi** [comment]
@yao-matrix

**Comment by sywangyi** [comment]
@danieldk please help review the PR. thanks

**Comment by ErikKaum** [comment]
Hi @sywangyi 👋 

Sorry that it's taken so long for anyone to respond. We'll try to get to this as fast as possible 🙌

File: server/text_generation_server/models/flash_causal_lm.py [hunk 1]
```diff
@@ -74,6 +74,36 @@ def get_sliding_windows() -> int:
...
+    import importlib.util
+
+    if importlib.util.find_spec("numa") is not None:
+        import numa
+        import psutil
+
+        nodes = numa.get_max_node() + 1
+        rank_per_node = math.ceil(world_size / nodes)
+        num_cpus_per_nodes = int(psutil.cpu_count(logical=False) / nodes)
+        node_id = int(rank_id / rank_per_node)
```
Comment (rbrugaro) [inline server/text_generation_server/models/flash_causal_lm.py:87]: @sywangyi node_id = int((rank_id +1)/ rank_per_node)

**Pull request merged**